stumpy/stump_tiles.py

```diff
@@ -0,0 +1,794 @@
+    sliding window dot product found in the original STOMP method.
+    """
+    m_inverse = 1.0 / m
+    constant = (m - 1) * m_inverse * m_inverse  # (m - 1)/(m * m)
+    uint64_m = np.uint64(m)
+
+    iter_ranges = np.zeros((config.STUMPY_N_DIAGONALS, 3), dtype=np.int64)
+    covariances = np.zeros(config.STUMPY_N_DIAGONALS, dtype=np.float64)
+
+    for tile_idx in range(tiles_ranges[thread_idx, 0], tiles_ranges[thread_idx, 1]):
+        (
+            i_offset,  # i location of current tile in distance profile
+            j_offset,  # j location of current tile in distance profile
+            tile_height,  # height of current tile
+            tile_width,  # width of current tile
+            tile_lower_diag,  # index of first diagonal to traverse on current tile
+            tile_upper_diag,  # index of last diagonal to traverse on current tile + 1
+            _,
+        ) = tiles[tile_idx]
+
+        # total number of diagonals to traverse
+        n_diags = tile_upper_diag - tile_lower_diag
+        # number of diagonal "chunks" if we divide our diagonals into chunks
+        n_chunks = int(np.ceil(n_diags / config.STUMPY_N_DIAGONALS))
+
+        # chunk_idx is the index of the chunk we are dealing with
+        for chunk_idx in range(0, n_chunks):
+            # number of diagonals to traverse together
+            chunk_size = min(
+                config.STUMPY_N_DIAGONALS,
+                n_diags - (chunk_idx * config.STUMPY_N_DIAGONALS),
+            )
+
+            # the longest diagonal to be traversed within the current chunk of diagonals
+            longest_diag_len = 0
+            # chunk_diag_idx is the diagonal index we are dealing with,
+            # with respect to the chunk
+            for chunk_diag_idx in range(chunk_size):
+                # tile_diag_idx is the diagonal index we are dealing with,
+                # with respect to the entire tile
+                tile_diag_idx = (
+                    tile_lower_diag
+                    + (chunk_idx * config.STUMPY_N_DIAGONALS)
+                    + chunk_diag_idx
+                )
+
+                # here we store the diagonal index we are dealing with,
+                # with respect to the entire distance matrix
+                iter_ranges[chunk_diag_idx, 0] = tile_diag_idx - i_offset + j_offset
+                # here we store the i location where the current diagonal begins
+                iter_ranges[chunk_diag_idx, 1] = i_offset - min(0, tile_diag_idx)
+                # here we store the i location where the current diagonal ends
+                iter_ranges[chunk_diag_idx, 2] = i_offset + min(
+                    tile_height, tile_width - tile_diag_idx
+                )
+
+                # record length of longest diagonal
+                longest_diag_len = max(
+                    longest_diag_len,
+                    iter_ranges[chunk_diag_idx, 2] - iter_ranges[chunk_diag_idx, 1],
+                )
+
```

Comment:
Now, I am thinking about this for-loop above. Can we get rid of it? I am trying to vectorize it:
```
tile_lower_diag_in_chunk = tile_lower_diag + chunk_idx * config.STUMPY_N_DIAGONALS
tile_diag_idx = tile_lower_diag_in_chunk + range(chunk_size) # instead of `for .. in range(chunk_size)`
iter_ranges[tile_diag_idx, 0] = tile_diag_idx + diag_inverse_shift ()
iter_ranges[chunk_diag_idx, 1] = i_offset - np.minimum(0, tile_diag_idx) # or maybe np.clip(...)
iter_ranges[chunk_diag_idx, 2] = ... 

longest_diag_len = np.max(iter_ranges[chunk_diag_idx, 2])
```

----

Also, instead of `range(chunk_size)`, we may do:
```
chunk_maxsize_range = np.arange(config.STUMPY_N_DIAGONALS) 
```
outside of the most outter loop, and then replace `range(chunk_size)` with `chunk_maxsize_range[:chunk_size]`

Reply:
This was an excellent suggestion, I hadn't considered this! Adopted!